### Importing needed libraries

In [1]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn import metrics
from sklearn.svm import SVC
from sklearn import metrics
from sklearn.tree import DecisionTreeClassifier
import xgboost as xgb

In [2]:
#I had to use the try and except block to catch the exception

import json
from itertools import count
reviews = []
with open(r'C:\Users\Stina\Documents\LEARNING PYTHON\My data for data science\All_Beauty_5.json') as f:

    for i in f:
        revf = json.loads(i)
        try:
            reviews.append((revf['reviewText'],revf['overall']))
        except KeyError:
            pass


In [3]:
#This function helps to get the sentiment score of each score

def get_sentiment(score):
    if score <= 2:
        return 'Negative'
    elif score == 3:
        return 'Neutral'
    elif score >3:
        return 'Positive'
        

In [4]:
#Getting the x and y values
# x is the reviews
# y is the sentiment

y=[get_sentiment(reviews[i][1]) for i in range(len(reviews))]
x=[reviews[i][0] for i in range(len(reviews))]

## Bag of words vectorization

In [5]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.25,random_state=True,shuffle=True)
vect = CountVectorizer()
vect_words=vect.fit_transform(x_train) #This learns the vocabulary of the data and transforms it into an array

#This transforms the x_test data into an array using the vocabulary learnt from the x_train
test_x = vect.transform(x_test) 

In [21]:
vect_words

<3948x5381 sparse matrix of type '<class 'numpy.int64'>'
	with 105593 stored elements in Compressed Sparse Row format>

In [6]:
#turning the y_data to a numeric data

from sklearn import preprocessing
le=preprocessing.LabelEncoder()
y_train = le.fit_transform(y_train)
y_test = le.fit_transform(y_test)



## Implementing a model


In [7]:
clf_dec = DecisionTreeClassifier()
clf_dec.fit(vect_words,y_train)
y_pred=clf_dec.predict(test_x)
clf_dec.score(test_x,y_test)  #This gives the accuracy of the model
print(metrics.classification_report(y_test,y_pred)) #This gives the report of all
metrics.accuracy_score(y_test,y_pred)  #This gives the accuracy of the model



              precision    recall  f1-score   support

           0       0.89      0.91      0.90        43
           1       0.70      0.61      0.65        23
           2       0.99      0.99      0.99      1250

    accuracy                           0.98      1316
   macro avg       0.86      0.84      0.85      1316
weighted avg       0.98      0.98      0.98      1316



0.9825227963525835

In [8]:
clf_svm = SVC(kernel = 'linear',degree = 2)
clf_svm.fit(vect_words,y_train)
y_pred=clf_svm.predict(test_x)
print(clf_svm.score(test_x,y_test)) #This gives the accuracy of the model
print(metrics.classification_report(y_test,y_pred)) #This gives the report of all metrics
print(metrics.accuracy_score(y_test,y_pred)) #This gives the accuracy of the model

0.986322188449848
              precision    recall  f1-score   support

           0       0.91      0.91      0.91        43
           1       0.83      0.65      0.73        23
           2       0.99      1.00      0.99      1250

    accuracy                           0.99      1316
   macro avg       0.91      0.85      0.88      1316
weighted avg       0.99      0.99      0.99      1316

0.986322188449848


In [9]:
xgbb = xgb.XGBClassifier(random_state=5,learning_rate=0.1,n_estimators=1000)

In [10]:
mode=xgbb.fit(vect_words,y_train)

In [11]:
metrics.accuracy_score(y_test,mode.predict(test_x))

0.9893617021276596

##### I tried 3 models and a xgboost boosting algorithm gave the highest prediction of 0.989